File: app/src/main/java/ru/melodin/fast/api/method/MethodSetter.java
package ru.melodin.fast.api.method;

import androidx.annotation.Nullable;
import androidx.collection.ArrayMap;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Collection;

import ru.melodin.fast.api.UserConfig;
import ru.melodin.fast.api.VKApi;
import ru.melodin.fast.api.model.VKModel;
import ru.melodin.fast.util.ArrayUtil;

public class MethodSetter {
    private String name;
    private ArrayMap<String, String> params;

    public MethodSetter(String name) {
        this.name = name;
        this.params = new ArrayMap<>();
    }

    public MethodSetter put(String key, Object value) {
        this.params.put(key, String.valueOf(value));
        return this;
    }

    public MethodSetter put(String key, int value) {
        this.params.put(key, String.valueOf(value));
        return this;
    }

    public MethodSetter put(String key, long value) {
        this.params.put(key, String.valueOf(value));
        return this;
    }

    public MethodSetter put(String key, boolean value) {
        this.params.put(key, value ? "1" : "0");
        return this;
    }

    private String getSignedUrl() {
        return getSignedUrl(false);
    }

    private String getSignedUrl(boolean isPost) {
        if (!params.containsKey("access_token")) {
            params.put("access_token", UserConfig.accessToken);
        }
        if (!params.containsKey("v")) {
            params.put("v", VKApi.API_VERSION);
        }
        if (!params.containsKey("lang")) {
            params.put("lang", VKApi.lang);
        }

        return VKApi.BASE_URL + name + "?" + (isPost ? "" : getParams());
    }

    public String getParams() {
        if (params == null || params.size() == 0) return "";

        StringBuilder buffer = new StringBuilder();
        try {

            for (int i = 0; i < params.size(); i++) {
                String key = params.keyAt(i);
                String value = params.valueAt(i);

                if (buffer.length() != 0) {
                    buffer.append("&");
                }

                buffer.append(key)
                        .append("=")
                        .append(URLEncoder.encode(value, "UTF-8"));
            }
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
        }
        return buffer.toString();
    }

    public void execute() throws Exception {
        VKApi.execute(getSignedUrl(), null);
    }

    public <E> ArrayList<E> execute(@Nullable Class<E> cls) throws Exception {
        return VKApi.execute(getSignedUrl(), cls);
    }

    public <E> void execute(@Nullable Class<E> cls, VKApi.OnResponseListener<E> listener) {
        VKApi.execute(getSignedUrl(), cls, listener);
    }

    public <E extends VKModel> ArrayList<E> tryExecute(@Nullable Class<E> cls) {
        try {
            return execute(cls);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    public MethodSetter userId(int value) {
        return put("user_id", value);
    }

    public MethodSetter userIds(Integer... ids) {
        return put("user_ids", ArrayUtil.toString(ids));
    }

    public MethodSetter userIds(Collection<Integer> ids) {
        return put("user_ids", ArrayUtil.toString(ids.toArray()));
    }

    public MethodSetter ownerId(int value) {
        return put("owner_id", value);
    }

    public MethodSetter peerId(int value) {
        return put("peer_id", value);
    }

    public MethodSetter groupId(int value) {
        return put("group_id", value);
    }

    public MethodSetter groupIds(int... ids) {
        return put("group_ids", ArrayUtil.toString(ids));
    }

    public MethodSetter fields(String values) {
        return put("fields", values);
    }

    public MethodSetter count(int value) {
        return put("count", value);
    }

    public MethodSetter sound(boolean value) {
        return put("sound", value);
    }

    public MethodSetter sort(int value) {
        return put("sort", value);
    }

    public MethodSetter time(int value) {
        return put("time", value);
    }

    public MethodSetter order(String value) {
        return put("order", value);
    }

    public MethodSetter offset(int value) {
        return put("offset", value);
    }

    public MethodSetter nameCase(String value) {
        return put("name_case", value);
    }

    public MethodSetter captchaSid(String value) {
        return put("captcha_sid", value);
    }

    public MethodSetter captchaKey(String value) {
        return put("captcha_key", value);
    }

    public MethodSetter withConfig(UserConfig config) {
        return put("access_token", UserConfig.accessToken);
    }

    public MethodSetter extended(boolean value) {
        return put("extended", value);
    }
}

File: app/src/main/java/ru/melodin/fast/api/method/MethodSetter.kt
package ru.melodin.fast.api.method

import androidx.collection.ArrayMap
import ru.melodin.fast.api.OnCompleteListener
import ru.melodin.fast.api.UserConfig
import ru.melodin.fast.api.VKApi
import ru.melodin.fast.util.ArrayUtil
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.*

open class MethodSetter(private val name: String) {
    private val params: ArrayMap<String, String>?

    private val signedUrl: String
        get() = getSignedUrl(false)

    init {
        this.params = ArrayMap()
    }

    fun put(key: String, value: Any?): MethodSetter {
        this.params!![key] = value.toString()
        return this
    }

    fun put(key: String, value: Int): MethodSetter {
        this.params!![key] = value.toString()
        return this
    }

    fun put(key: String, value: Long): MethodSetter {
        this.params!![key] = value.toString()
        return this
    }

    fun put(key: String, value: Boolean): MethodSetter {
        this.params!![key] = if (value) "1" else "0"
        return this
    }

    private fun getSignedUrl(isPost: Boolean): String {
        if (!params!!.containsKey("access_token")) {
            params["access_token"] = UserConfig.accessToken
        }
        if (!params.containsKey("v")) {
            params["v"] = VKApi.API_VERSION
        }
        if (!params.containsKey("lang")) {
            params["lang"] = VKApi.lang
        }

        return VKApi.BASE_URL + name + "?" + if (isPost) "" else getParams()
    }

    fun getParams(): String {
        if (ArrayUtil.isEmpty(params)) return ""

        params ?: return ""

        val buffer = StringBuilder()
        try {
            for (i in 0 until params.size) {
                val key = params.keyAt(i)
                val value = params.valueAt(i)

                if (buffer.isNotEmpty()) {
                    buffer.append("&")
                }

                buffer.append(key)
                        .append("=")
                        .append(URLEncoder.encode(value, "UTF-8"))
            }
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
        }

        return buffer.toString()
    }

    @Throws(Exception::class)
    fun execute() {
        VKApi.execute<Any>(signedUrl, null)
    }

    @Throws(Exception::class)
    fun execute(cls: Class<*>?): ArrayList<*>? {
        return VKApi.execute(signedUrl, cls)
    }

    fun execute(cls: Class<*>?, listener: OnCompleteListener) {
        VKApi.execute(signedUrl, cls, listener)
    }

    fun userId(value: Int): MethodSetter {
        return put("user_id", value)
    }

    fun userIds(vararg ids: Int): MethodSetter {
        return put("user_ids", ArrayUtil.toString(*ids))
    }

    fun userIds(ids: Collection<Int>): MethodSetter {
        return put("user_ids", ArrayUtil.toString(ids))
    }

    fun ownerId(value: Int): MethodSetter {
        return put("owner_id", value)
    }

    fun peerId(value: Int): MethodSetter {
        return put("peer_id", value)
    }

    fun groupId(value: Int): MethodSetter {
        return put("group_id", value)
    }

    fun groupIds(vararg ids: Int): MethodSetter {
        return put("group_ids", ArrayUtil.toString(*ids))
    }

    fun fields(values: String): MethodSetter {
        return put("fields", values)
    }

    fun count(value: Int): MethodSetter {
        return put("count", value)
    }

    fun sound(value: Boolean): MethodSetter {
        return put("sound", value)
    }

    fun sort(value: Int): MethodSetter {
        return put("sort", value)
    }

    fun time(value: Int): MethodSetter {
        return put("time", value)
    }

    fun order(value: String): MethodSetter {
        return put("order", value)
    }

    fun offset(value: Int): MethodSetter {
        return put("offset", value)
    }

    fun nameCase(value: String): MethodSetter {
        return put("name_case", value)
    }

    fun captchaSid(value: String): MethodSetter {
        return put("captcha_sid", value)
    }

    fun captchaKey(value: String): MethodSetter {
        return put("captcha_key", value)
    }

    fun withConfig(config: UserConfig): MethodSetter {
        return put("access_token", UserConfig.accessToken)
    }

    open fun extended(value: Boolean): MethodSetter {
        return put("extended", value)
    }
}
